agnpy/time_evolution/time_evolution.py

```diff
@@ -0,0 +1,442 @@
+        gm_bins_up = np.zeros_like(gm_bins_lb)
+        gm_bins = np.array([gm_bins_lb, gm_bins_up])
+        update_bin_ub(gm_bins)
+        density = self._blob.n_e(gm_bins_lb)
+        if self._subgroups is not None:
+            if self._subgroups_initial_density is not None:
+                subgroups_density = self._subgroups_initial_density
+            else:
+                subgroups_density = np.zeros((len(self._subgroups), len(self._initial_gamma_array)), dtype=float)
+                subgroups_density[0] = np.ones(len(self._initial_gamma_array), dtype=float)
+        else:
+            subgroups_density = np.ones((1, len(self._initial_gamma_array)), dtype=float)
+        en_chg_rates = recalc_new_rates(gm_bins, self._energy_change_functions, erg_per_s, density, subgroups_density)
+        rel_injection_rates = recalc_new_rates(gm_bins_lb, self._injection_functions_rel, per_s, density, subgroups_density)
+        abs_injection_rates = recalc_new_rates(gm_bins_lb, self._injection_functions_abs, per_s_cm3, density, subgroups_density)
+        return gm_bins, density, subgroups_density, en_chg_rates, abs_injection_rates, rel_injection_rates
+
+    def evaluate(self) -> TimeEvaluationResult:
+        """
+        Performs the time evolution of the electron distribution inside the blob. In the loop, for each gamma point it creates a narrow bin,
+        calculates the energy change for a start and an end of the bin, and scales up density by the bin narrowing factor and injection/escape rate.
+
+        Side Effects
+        ------------
+        Replaces the blob.n_e with the new InterpolatedDistribution
+
+        Returns
+        ------------
+        A tuple consisting of final values of: gamma, densities, density groups, change rates, absolute injection rates, relative injection rates
+        """
+
+        gm_bins, density, subgroups_density, en_chg_rates, abs_injection_rates, rel_injection_rates = self._calculate_initial_values()
+        total_time_elapsed = 0 * u.s
+        log.info("Progress: %3d%% %s (%i bins)", 0, total_time_elapsed, gm_bins.shape[-1])
+        time_left_per_bin = np.zeros_like(gm_bins[0]) * u.s
+
+        while total_time_elapsed < self._total_time_sec:
+            en_bins = to_erg(gm_bins)
+            en_chg_rates_grouped, rel_injection_rates_grouped, abs_injection_rates_grouped = self._group_change_rates(
+                gm_bins, en_chg_rates, rel_injection_rates, abs_injection_rates)
+            densities_grouped = to_densities_grouped(density, subgroups_density)
+            total_injection_rates_grouped = densities_grouped * rel_injection_rates_grouped + abs_injection_rates_grouped
+            if self._step_duration != "auto":
+                step_time = self._step_duration
+            else:
+                max_times = self._calc_max_time_per_bin(
+                    en_bins, density, subgroups_density, en_chg_rates_grouped, rel_injection_rates_grouped, abs_injection_rates_grouped)
```

Comment:
if you have a function with so many parameters it is more clear and less error prone not to use position arguments but execute it in this style ...(AAA=aaa, BBB=bbb, ...)
especially that some of the functions take first rel_inejction and some abs_injection